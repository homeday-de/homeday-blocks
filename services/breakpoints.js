let breakpoints = {};

export const matchMediaAvailable = (
  typeof window !== 'undefined'
  && typeof window.matchMedia === 'function'
);

export const mediaMatches = (breakpoint, {
  matchMedia = matchMediaAvailable ? window.matchMedia : () => false,
} = {}) => {
  if (typeof breakpoints[breakpoint] === 'undefined') {
    // Breakpoint not defined, we treat it as a media query string
    return matchMedia(breakpoint).matches;
  }

  return matchMedia(breakpoints[breakpoint]).matches;
};

export const setBreakpoints = (newBreakpoints) => {
  breakpoints = newBreakpoints;
};

export const getBreakpoints = () => breakpoints;

export default {
  matchMediaAvailable,
  mediaMatches,
  setBreakpoints,
  getBreakpoints,
};
