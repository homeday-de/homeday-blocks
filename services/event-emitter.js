import Vue from 'vue';

const EventEmitter = new Vue();

export default EventEmitter;
