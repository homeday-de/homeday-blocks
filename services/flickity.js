export function hidePaginationWhenNotNeeded() {
  const viewport = this.size.width;
  const fullSize = Math.round(this.slideableWidth);

  if (fullSize > viewport) {
    this.bindDrag();
    this.element.classList.remove('flickity--no-controls');
  } else if (fullSize <= viewport) {
    this.unbindDrag();
    this.element.classList.add('flickity--no-controls');
  }
}

export default {
  hidePaginationWhenNotNeeded,
};
