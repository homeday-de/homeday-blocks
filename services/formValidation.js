export function email(string) {
  // eslint-disable-next-line
  const regex = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
  return regex.test(string.toLowerCase());
}

export function date(string) {
  // eslint-disable-next-line
  const regex = /[0-9]{4}-[0-9]{2}-[0-9]{2}/;
  return regex.test(string);
}

export default {
  email,
  date,
};
