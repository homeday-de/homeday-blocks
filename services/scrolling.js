import naturalScroll from 'natural-scroll';
import {
  disableBodyScroll,
  enableBodyScroll,
} from 'body-scroll-lock';

export function elOffset(el) {
  const box = el.getBoundingClientRect();
  const clientTop = document.clientTop || document.body.clientTop || 0;
  const clientLeft = document.clientLeft || document.body.clientLeft || 0;
  const scrollTop = window.pageYOffset || document.scrollTop || 0;
  const scrollLeft = window.pageXOffset || document.scrollLeft || 0;

  return {
    top: box.top + (scrollTop - clientTop),
    left: box.left + (scrollLeft - clientLeft),
    width: box.width,
    height: box.height,
  };
}

export function getDocumentHeight() {
  return Math.max(
    document.body.scrollHeight, document.documentElement.scrollHeight,
    document.body.offsetHeight, document.documentElement.offsetHeight,
    document.body.clientHeight, document.documentElement.clientHeight,
  );
}

export function isWindowFullyScrolled() {
  if (window.pageYOffset + document.documentElement.clientHeight >= getDocumentHeight()) {
    return true;
  }

  return false;
}

export function freezeScrolling(el) {
  // eslint-disable-next-line no-param-reassign
  el.style['-webkit-overflow-scrolling'] = 'touch';
  disableBodyScroll(el);
}

export function unfreezeScrolling(el) {
  // eslint-disable-next-line no-param-reassign
  el.style['-webkit-overflow-scrolling'] = 'auto';
  enableBodyScroll(el);
}

export function isElVisible({ el, offset = 0 }) {
  if (el == null) {
    return false;
  }

  const { top: offsetTop } = elOffset(el);

  if (offsetTop < window.pageYOffset + offset) {
    return false;
  }

  if (offsetTop > window.pageYOffset + window.innerHeight) {
    return false;
  }

  return true;
}

export function isScrolledPastElBottom({
  el,
  // The default offset will check whether we reached the bottom of the element
  // with the bottom edge of the screen
  offset = window.innerHeight,
}) {
  if (el == null) {
    return false;
  }

  const { top: offsetTop } = elOffset(el);

  const elBottom = offsetTop + el.clientHeight;

  if (elBottom > window.pageYOffset + offset) {
    return false;
  }

  return true;
}

export function scrollToEl({ el, offset = 0, onlyIfNotVisible = true }) {
  const toScroll = [
    document.documentElement,
    document.body,
  ];

  if (el == null || (onlyIfNotVisible && isElVisible({ el, offset }))) {
    return;
  }

  toScroll.forEach((toScrollEl) => {
    naturalScroll.scrollTop(toScrollEl, elOffset(el).top - offset);
  });
}

export default {
  elOffset,
  getDocumentHeight,
  isWindowFullyScrolled,
  freezeScrolling,
  unfreezeScrolling,
  isElVisible,
  isScrolledPastElBottom,
  scrollToEl,
};
