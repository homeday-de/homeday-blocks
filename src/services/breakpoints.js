import sassVariables from 'homeday-blocks/src/styles/variablesForJS.scss';

let breakpoints = {
  mobile: `(max-width: ${sassVariables.breakMobile - 1}px)`,
  tablet: `(min-width: ${sassVariables.breakMobile}px)`,
  desktop: `(min-width: ${sassVariables.breakTablet}px)`,
  desktopWide: `(min-width: ${sassVariables.breakDesktop}px)`,
  desktopExtraWide: `(min-width: ${sassVariables.breakDesktopWide}px)`,
};

export const matchMediaAvailable = window && typeof window.matchMedia === 'function';

const defaultMatchMedia = matchMediaAvailable ? window.matchMedia : () => false;

export function mediaMatches(
  breakpoint,
  { matchMedia = defaultMatchMedia } = {},
) {
  if (typeof breakpoints[breakpoint] === 'undefined') {
    // Breakpoint not defined, we treat it as a media query string
    return matchMedia(breakpoint).matches;
  }

  return matchMedia(breakpoints[breakpoint]).matches;
}

export function setBreakpoints(newBreakpoints) {
  breakpoints = newBreakpoints;
}

export function getBreakpoints() {
  return breakpoints;
}

export default {
  matchMediaAvailable,
  mediaMatches,
  setBreakpoints,
  getBreakpoints,
};
