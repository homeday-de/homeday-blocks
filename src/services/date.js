// Sets date time to midnight for unifying purposes
export const resetDateTime = (date) => {
  const clonedDate = new Date(date);
  clonedDate.setHours(0, 0, 0, 0);

  return clonedDate;
};

// Returns date which is amountOfDays from the initialDate
export const getNDaysFromDate = (initialDate, amountOfDays) => {
  // Create a clone of initialDate in order not to edit it
  const startDate = new Date(initialDate);
  return resetDateTime(new Date(startDate.setDate(initialDate.getDate() + amountOfDays)));
};

export const convertMsToDays = ms => Math.round(ms / 1000 / 60 / 60 / 24);

export const getDaysDateRange = (startDate, endDate) => {
  const dateDiffDays = convertMsToDays(endDate.getTime() - startDate.getTime());
  const datesArray = [];

  /* eslint-disable no-plusplus  */
  for (let day = 0; day < dateDiffDays; day++) {
    datesArray.push(getNDaysFromDate(startDate, day));
  }

  return datesArray;
};

// Returns array of unique date strings from original array
export const getIntlDateString = (locale, datesArray, localStringArgs) => datesArray.reduce((accumulator, currentItem) => {
  const currentDay = new Date(currentItem).toLocaleString(
    locale,
    localStringArgs,
  );
  if (accumulator.indexOf(currentDay) === -1) accumulator.push(currentDay);

  return accumulator;
}, []);

export const generateDateCycles = (cycleLengthWeeks, amountOfCycles, startDateOffset) => {
  const WEEK_DAYS = 7;
  const cycleLengthDays = cycleLengthWeeks * WEEK_DAYS;
  const startDate = getNDaysFromDate(new Date(), startDateOffset);
  const endDate = getNDaysFromDate(new Date(), (cycleLengthDays * amountOfCycles) + startDateOffset);

  return getDaysDateRange(startDate, endDate);
};
