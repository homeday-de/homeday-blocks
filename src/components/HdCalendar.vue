<template>
  <div class="calendar">
    <header class="calendar__header">
      <button
        class="calendar__button calendar__header__button calendar__button--prev"
        :class="{'calendar__button--disabled': selectedWeekIndexTail - cycleLengthWeeks < 0}"
        @click="setWeekIndexTail(-cycleLengthWeeks)"
      />
      <span class="calendar__header__month">
        <template v-for="(month, index) in activeMonths">
          {{ index > 0 ? `- ${month}` : month }}
        </template>
      </span>
      <button
        class="calendar__button calendar__header__button calendar__button--next"
        :class="{'calendar__button--disabled': selectedWeekIndexTail + cycleLengthWeeks >= availableWeeks.length}"
        @click="setWeekIndexTail(cycleLengthWeeks)"
      />
    </header>
    <section class="calendar__body">
      <header class="calendar__days calendar__row">
        <div v-for="day in weekdays" :key="day" class="calendar__row__item calendar__day">{{day}}</div>
      </header>
      <div class="calendar__body__inner">
        <transition :name="'calendar-' + animationDirection">
          <section class="calendar__dates" v-if="animateCalendar">
            <div
              class="calendar__week calendar__row"
              v-for="(displayWeek, index) in displayedWeeks"
              :key="index"
            >
              <div
                v-for="(date, index) in displayWeek"
                class="calendar__date calendar__row__item"
                :class="{'calendar__date--disabled': isDisabled(date, index), 'calendar__date--selected': isSelectedDate(date)}"
                :key="date.getDate()"
                @click="selectDate(date)"
                @touchstart="selectDate(date)"
              >
                {{ date.getDate() }}
              </div>
            </div>
          </section>
        </transition>
      </div>
    </section>
  </div>
</template>


<script>
import flatten from 'lodash/flatten';
import chunk from 'lodash/chunk';
import { getIntlDateString } from 'hd-blocks/services/date';

const WEEK_DAYS = 7;
export default {
  name: 'hdCalendar',
  props: {
    dates: {
      type: Array,
      required: true,
      // Check if passed array contains only valid dates
      validator: dates => dates.filter(d => typeof d.getMonth === 'function').length === dates.length,
    },
    disabledIndexes: {
      type: Array,
      default() {
        return [];
      },
    },
    cycleLengthWeeks: {
      type: Number,
      default: 2,
    },
    locale: {
      type: String,
      default: 'de',
    },
  },
  data: () => ({
    weekdays: [],
    availableWeeks: [],
    selectedDate: null,
    selectedWeekIndexTail: 0,
    animateCalendar: true,
    animationDirection: 'right',
  }),
  mounted() {
    this.weekdays = getIntlDateString(this.locale, this.dates, { weekday: 'short' });
    // Slice dates array into week chunks
    this.availableWeeks = chunk(this.dates, WEEK_DAYS);

    this.selectFirstAvialable();
  },
  computed: {
    displayedWeeks() {
      return this.availableWeeks.slice(this.selectedWeekIndexTail, this.selectedWeekIndexTail + this.cycleLengthWeeks);
    },
    activeMonths() {
      return getIntlDateString(this.locale, flatten(this.displayedWeeks), { month: 'long' });
    },
  },
  watch: {
    disabledIndexes() {
      this.selectFirstAvialable();
    },
  },
  methods: {
    selectFirstAvialable() {
      const firstAvialableDay = this.displayedWeeks[0]
        .find(day => this.isDisabled(day) === false);

      this.selectDate(firstAvialableDay);
    },
    isSelectedDate(date) {
      return this.selectedDate ? this.selectedDate.toUTCString() === date.toUTCString() : false;
    },
    setWeekIndexTail(indexDiff) {
      const maxBound = this.selectedWeekIndexTail + indexDiff >= this.availableWeeks.length;
      const minBound = this.selectedWeekIndexTail + indexDiff < 0;

      this.animationDirection = indexDiff > 0 ? 'right' : 'left';
      if (maxBound || minBound) return;

      this.animateCalendar = !this.animateCalendar;
      setTimeout(() => {
        this.selectedWeekIndexTail += indexDiff;
        this.animateCalendar = !this.animateCalendar;
        this.selectFirstAvialable();
      }, 0);
    },
    isDisabled(date) {
      const dateIndex = this.dates.indexOf(date);
      const isDisabledIndex = this.disabledIndexes.indexOf(dateIndex) !== -1;

      return isDisabledIndex || this.isWeekend(date);
    },
    isWeekend(date) {
      // IE's Date.toLocaleString is includes left-to-right marks in it (U+200E), so we have to trim that
      const dateDayString = date.toLocaleString('en-GB', { weekday: 'short' }).replace(/[^ -~]/g, '');
      return dateDayString === 'Sun' || dateDayString === 'Sat';
    },
    selectDate(date) {
      if (this.isDisabled(date)) return;
      const dateIndex = this.dates.findIndex(d => d === date);
      this.selectedDate = date;
      this.$emit('dateSelected', {
        date,
        dateIndex,
      });
    },
  },
};
</script>

<style lang="scss" scoped>
@import 'hd-blocks/styles/mixins.scss';

// Transitions
.calendar {
  &-left {
    &-leave-active, &-enter-active {
      transition: transform 200ms ease;
    }
    &-enter, &-leave-to {
      transform: translate(-100%, 0);
      opacity: 0;
    }
  }
  &-right {
    &-leave-active, &-enter-active {
      transition: transform 200ms ease;
    }
    &-enter, &-leave-to {
      transform: translate(100%, 0);
      opacity: 0;
    }
  }
}

.calendar {
  width: 100%;
  border-bottom: 1px solid $heather;

  &__header {
    text-align: center;
    position: relative;
    margin-bottom: $stack-m;
    display: flex;
    justify-content: center;
    align-items: center;
    &__month {
      font-size: 14px;
      line-height: 20px;
      font-weight: bold;
      margin: 0 $stack-xl;

      @media (min-width: $break-mobile) {
        font-size: 18px;
        line-height: 23px;
      }
    }

    @media (min-width: $break-mobile) {
      margin-bottom: $stack-s * 3;
    }
  }
  &__row {
    display: flex;
    margin-bottom: $stack-m;

    @media (min-width: $break-mobile) {
      margin-bottom: $stack-m;
    }

    &__item {
      width: calc(100% / 7);
      text-align: center;
      &:not(:last-child) {
        margin-right: $inline-m;
      }
    }
  }
  &__days {
    margin-bottom: $stack-m;

    @media (min-width: $break-mobile) {
      margin-bottom: $stack-s * 3;
    }
  }
  &__day {
    font-size: 14px;
    line-height: 10px;
    color: $regent-gray;
    text-transform: uppercase;
    font-weight: 600;
    @media (min-width: $break-mobile) {
      font-size: 18px;
    }
  }

  &__date {
    position: relative;
    font-size: 16px;
    z-index: 1;
    cursor: pointer;

    @media (min-width: $break-mobile) {
      font-size: 18px;
    }
    &--disabled {
      color: $heather;
      cursor: not-allowed;
    }
    &:after {
      content: "";
      width: 16px;
      height: 16px;
      background: $vivid-blue;
      position: absolute;
      left: 0;
      right: 0;
      margin: auto;
      top: 0;
      bottom: 0;
      z-index: -1;
      border-radius: 50%;
      transform: scale(0);
      transition: transform ease 200ms;
    }


    &:not(&--disabled):hover {
      &:after {
        transform: scale(0.3) translateY(250%);
      }
    }
    :root &--selected:not(&--disabled) {
      cursor: initial;
      color: $white;

      &:after {
        transform: scale(2) translateY(0);
        @media (min-width: $break-mobile) {
          transform: scale(3) translateY(0);
        }
      }
    }
  }
  &__dates {
    width: 100%;
  }
  &__button {
    border: none;
    width: 16px;
    height: 16px;
    cursor: pointer;
    z-index: 2;

    @media (min-width: $break-mobile) {
      display: block;
      width: 24px;
      height: 24px;
    }
    &--disabled {
      opacity: 0.4;
      cursor: not-allowed;
    }
    &--prev {
      background: url('~hd-blocks/assets/icons/ic_arrow_backwards--idle-24px.svg') no-repeat center;
    }
    &--next {
      background: url('~hd-blocks/assets/icons/ic_arrow_forward--idle-24px.svg') no-repeat center;
    }

    &--mobile {
      display: flex;
    }
  }
  &__body {
    &__inner {
      width: 100%;
      display: flex;
      justify-content: center;
      position: relative;
    }
    &__button {
      display: none;
      position: absolute;
      top: 0;
      bottom: 0;
      margin: auto;

      &:first-child {
        left: 0;
      }
    }
  }
}
</style>
