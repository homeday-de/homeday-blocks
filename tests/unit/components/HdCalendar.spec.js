import { wrapperFactoryBuilder } from 'tests/unit/helpers';
import { generateDateCycles } from '@/services/date';
import HdCalendar from '@/components/HdCalendar.vue';

const wrapperBuilder = wrapperFactoryBuilder(HdCalendar, {
  props: {
    dates: generateDateCycles(2, 2, 2, new Date(1789, 4, 5)),
  },
  shallow: true,
});

describe('HdCalendar', () => {
  it('renders component', async () => {
    const wrapper = wrapperBuilder();

    await wrapper.vm.$nextTick();

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('as a user I\'m able to change week', () => {
    const setWeekIndexTailMock = jest.fn();
    const wrapper = wrapperBuilder({
      methods: {
        setWeekIndexTail: setWeekIndexTailMock,
      },
    });

    wrapper.find('.calendar__button--next').trigger('click');
    expect(setWeekIndexTailMock).toHaveBeenCalledTimes(1);

    wrapper.find('.calendar__button--prev').trigger('click');
    expect(setWeekIndexTailMock).toHaveBeenCalledTimes(2);
  });

  it('as a user I\'m able to change week and the first available day gets selected', () => {
    jest.useFakeTimers();

    const selectDateMock = jest.fn();
    const wrapper = wrapperBuilder({
      methods: {
        selectDate: selectDateMock,
      },
    });

    wrapper.find('.calendar__button--next').trigger('click');
    jest.advanceTimersByTime(0);

    expect(selectDateMock).toHaveBeenCalled();
  });

  it('as a user I\'m able to select a day', async () => {
    const wrapper = wrapperBuilder();

    await wrapper.vm.$nextTick();

    const firstDate = wrapper.find('.calendar__date');
    firstDate.trigger('click');

    expect(firstDate.classes()).toContain('calendar__date--selected');
  });

  it('when the disabled indexes are updated, the first available date is selected', () => {
    const selectFirstAvialableMock = jest.fn();
    const wrapper = wrapperBuilder({
      methods: {
        selectFirstAvialable: selectFirstAvialableMock,
      },
    });

    wrapper.setProps({
      disabledIndexes: [],
    });

    expect(selectFirstAvialableMock).toHaveBeenCalled();
  });
});
