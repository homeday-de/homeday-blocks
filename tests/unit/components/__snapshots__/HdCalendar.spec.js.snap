// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HdCalendar renders component 1`] = `
<div class="calendar">
  <header class="calendar__header"><button class="calendar__button calendar__header__button calendar__button--prev calendar__button--disabled">
      <!----></button> <span class="calendar__header__month">
        May
      </span> <button class="calendar__button calendar__header__button calendar__button--next">
      <!----></button></header>
  <section class="calendar__body">
    <header class="calendar__days calendar__row">
      <div class="calendar__row__item calendar__day">Thu</div>
      <div class="calendar__row__item calendar__day">Fri</div>
      <div class="calendar__row__item calendar__day">Sat</div>
      <div class="calendar__row__item calendar__day">Sun</div>
      <div class="calendar__row__item calendar__day">Mon</div>
      <div class="calendar__row__item calendar__day">Tue</div>
      <div class="calendar__row__item calendar__day">Wed</div>
    </header>
    <div class="calendar__body__inner">
      <section class="calendar__dates" name="calendar-right">
        <div class="calendar__week calendar__row">
          <div class="calendar__date calendar__row__item calendar__date--selected">
            7
          </div>
          <div class="calendar__date calendar__row__item">
            8
          </div>
          <div class="calendar__date calendar__row__item calendar__date--disabled">
            9
          </div>
          <div class="calendar__date calendar__row__item calendar__date--disabled">
            10
          </div>
          <div class="calendar__date calendar__row__item">
            11
          </div>
          <div class="calendar__date calendar__row__item">
            12
          </div>
          <div class="calendar__date calendar__row__item">
            13
          </div>
        </div>
        <div class="calendar__week calendar__row">
          <div class="calendar__date calendar__row__item">
            14
          </div>
          <div class="calendar__date calendar__row__item">
            15
          </div>
          <div class="calendar__date calendar__row__item calendar__date--disabled">
            16
          </div>
          <div class="calendar__date calendar__row__item calendar__date--disabled">
            17
          </div>
          <div class="calendar__date calendar__row__item">
            18
          </div>
          <div class="calendar__date calendar__row__item">
            19
          </div>
          <div class="calendar__date calendar__row__item">
            20
          </div>
        </div>
      </section>
    </div>
  </section>
</div>
`;
